In [1]:
import requests
import matplotlib.pyplot as plt
import json
import pandas as pd
from shapely.geometry import Point
import geopandas as gpd

In [2]:
endpoint = "https://data.nashville.gov/resource/2u6v-ujjs.json"

In [15]:
type(endpoint)

str

In [17]:
params={'offense_description': 'BURGLARY- AGGRAVATED', '$where': "incident_reported between '2022-01-01' and '2022-09-30'", '$limit': 50000}
response = requests.get(endpoint, params=params)
response

<Response [200]>

In [4]:
response.text

'[{"primary_key":"20220641257_11","incident_number":"20220641257","report_type":"W","report_type_description":"WITNESS","incident_status_code":"U","incident_status_description":"UNFOUNDED","investigation_status":"Closed","incident_occurred":"2022-10-14T16:00:00.000","incident_reported":"2022-10-14T16:01:00.000","incident_location":"4343 4343","latitude":"36.207","longitude":"-86.612","zip_code":"37076","location_code":"22","location_description":"RESIDENCE, HOME","offense_number":"1","offense_nibrs":"810","offense_description":"LOST PROPERTY","weapon_primary":"17","weapon_description":"NONE","victim_number":"1","domestic_related":false,"victim_type":"I","victim_description":"INDIVIDUAL (18 AND OVER)","victim_gender":"F","victim_race":"W","victim_ethnicity":"Non-Hispanic","victim_county_resident":"RESIDENT","mapped_location":{"type":"Point","coordinates":[-86.612,36.207]},":@computed_region_wvby_4s8j":"1",":@computed_region_3aw5_2wv7":"2",":@computed_region_p6sk_2acq":"27",":@computed_r

In [5]:
response

<Response [200]>

In [18]:
res = response.json()
res

[{'primary_key': '20220167824_11',
  'incident_number': '20220167824',
  'report_type': 'D',
  'report_type_description': 'DISPATCHED',
  'incident_status_code': 'O',
  'incident_status_description': 'OPEN',
  'investigation_status': 'Open',
  'incident_occurred': '2022-04-07T15:00:00.000',
  'incident_reported': '2022-04-08T11:52:00.000',
  'incident_location': 'JACKSON ST',
  'latitude': '36.17',
  'longitude': '-86.8',
  'location_code': '22',
  'location_description': 'RESIDENCE, HOME',
  'offense_number': '1',
  'offense_nibrs': '220',
  'offense_description': 'BURGLARY- AGGRAVATED',
  'weapon_primary': '16',
  'weapon_description': 'Unarmed',
  'victim_number': '1',
  'domestic_related': False,
  'victim_type': 'I',
  'victim_description': 'INDIVIDUAL (18 AND OVER)',
  'victim_gender': 'U',
  'victim_race': 'W',
  'victim_ethnicity': 'Non-Hispanic',
  'victim_county_resident': 'NON RESIDENT',
  'mapped_location': {'type': 'Point', 'coordinates': [-86.8, 36.17]}},
 {'primary_key':

In [19]:
burglaries = pd.DataFrame(res)
burglaries

primary_key incident_number report_type report_type_description  \
0     20220167824_11     20220167824           D              DISPATCHED   
1     20220126184_31     20220126184           D              DISPATCHED   
2     20220027854_12     20220027854           D              DISPATCHED   
3     20220032825_11     20220032825           D              DISPATCHED   
4     20220010799_11     20220010799           D              DISPATCHED   
...              ...             ...         ...                     ...   
1355  20220462314_11     20220462314           D              DISPATCHED   
1356  20220397543_11     20220397543           D              DISPATCHED   
1357  20220487005_11     20220487005           D              DISPATCHED   
1358  20220116140_31     20220116140           D              DISPATCHED   
1359  20220315247_11     20220315247           S                 SUSPECT   

     incident_status_code incident_status_description investigation_status  \
0                       O                        OPEN                 Open   
1                       O                        OPEN                 Open   
2                       O                        OPEN                 Open   
3                       O                        OPEN                 Open   
4                       O                        OPEN                 Open   
...                   ...                         ...                  ...   
1355                    R        REFUSED TO COOPERATE               Closed   
1356                    R        REFUSED TO COOPERATE               Closed   
1357                    R        REFUSED TO COOPERATE               Closed   
1358                    A           CLEARED BY ARREST               Closed   
1359                    O                        OPEN                 Open   

            incident_occurred        incident_reported incident_location  ...  \
0     2022-04-07T15:00:00.000  2022-04-08T11:52:00.000        JACKSON ST  ...   
1     2022-03-18T02:30:00.000  2022-03-18T06:51:00.000        BENTON AVE  ...   
2     2022-01-18T07:45:00.000  2022-01-19T23:48:00.000     CANE RIDGE RD  ...   
3     2022-01-23T00:40:00.000  2022-01-24T06:57:00.000     BROOKWOOD TER  ...   
4     2021-12-06T12:00:00.000  2022-01-21T16:25:00.000          SOUTH ST  ...   
...                       ...                      ...               ...  ...   
1355  2022-09-13T17:00:00.000  2022-09-15T02:04:00.000         1524 1524  ...   
1356  2022-08-11T18:00:00.000  2022-08-11T22:47:00.000           930 930  ...   
1357  2022-09-27T19:16:00.000  2022-09-27T22:46:00.000           400 400  ...   
1358  2022-03-13T01:49:00.000  2022-03-13T01:49:00.000           306 306  ...   
1359  2022-06-27T12:00:00.000  2022-06-27T19:46:00.000        HARDING PL  ...   

     victim_type        victim_description victim_gender victim_race  \
0              I  INDIVIDUAL (18 AND OVER)             U           W   
1              I  INDIVIDUAL (18 AND OVER)             M           W   
2              I  INDIVIDUAL (18 AND OVER)             M           B   
3              I  INDIVIDUAL (18 AND OVER)             F           W   
4              I  INDIVIDUAL (18 AND OVER)             M           W   
...          ...                       ...           ...         ...   
1355           I  INDIVIDUAL (18 AND OVER)             M           W   
1356           I  INDIVIDUAL (18 AND OVER)             F           B   
1357           I  INDIVIDUAL (18 AND OVER)             F           B   
1358           I  INDIVIDUAL (18 AND OVER)             F           B   
1359           I  INDIVIDUAL (18 AND OVER)             F           B   

     victim_ethnicity victim_county_resident  \
0        Non-Hispanic           NON RESIDENT   
1        Non-Hispanic               RESIDENT   
2        Non-Hispanic               RESIDENT   
3        Non-Hispanic               RESIDENT   
4        Non-Hispanic               RESIDENT   
...               ...            

In [8]:
burglaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 35 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   primary_key                  1000 non-null   object
 1   incident_number              1000 non-null   object
 2   report_type                  1000 non-null   object
 3   report_type_description      943 non-null    object
 4   incident_status_code         1000 non-null   object
 5   incident_status_description  1000 non-null   object
 6   investigation_status         1000 non-null   object
 7   incident_occurred            1000 non-null   object
 8   incident_reported            1000 non-null   object
 9   incident_location            969 non-null    object
 10  latitude                     466 non-null    object
 11  longitude                    466 non-null    object
 12  zip_code                     410 non-null    object
 13  location_code                1000 

In [9]:
burglaries["latitude"] = pd.to_numeric(burglaries["latitude"])
burglaries["longitude"] = pd.to_numeric(burglaries["longitude"])

In [10]:
burglaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 35 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   primary_key                  1000 non-null   object 
 1   incident_number              1000 non-null   object 
 2   report_type                  1000 non-null   object 
 3   report_type_description      943 non-null    object 
 4   incident_status_code         1000 non-null   object 
 5   incident_status_description  1000 non-null   object 
 6   investigation_status         1000 non-null   object 
 7   incident_occurred            1000 non-null   object 
 8   incident_reported            1000 non-null   object 
 9   incident_location            969 non-null    object 
 10  latitude                     466 non-null    float64
 11  longitude                    466 non-null    float64
 12  zip_code                     410 non-null    object 
 13  location_code      

In [11]:
burglaries["geometry"] = burglaries.apply(lambda x: Point((x['longitude'], x['latitude'])), axis=1)

C:\Users\hpunc\ANACONDA\lib\site-packages\pandas\core\dtypes\cast.py:122: ShapelyDeprecationWarning: The array interface is deprecated and will no longer work in Shapely 2.0. Convert the '.coords' to a numpy array instead.
  arr = construct_1d_object_array_from_listlike(values)


In [22]:
burglaries["geometry"] = gpd.points_from_xy(burglaries["longitude"], burglaries["latitude"])

In [23]:
burglaries

primary_key incident_number report_type report_type_description  \
0     20220167824_11     20220167824           D              DISPATCHED   
1     20220126184_31     20220126184           D              DISPATCHED   
2     20220027854_12     20220027854           D              DISPATCHED   
3     20220032825_11     20220032825           D              DISPATCHED   
4     20220010799_11     20220010799           D              DISPATCHED   
...              ...             ...         ...                     ...   
1355  20220462314_11     20220462314           D              DISPATCHED   
1356  20220397543_11     20220397543           D              DISPATCHED   
1357  20220487005_11     20220487005           D              DISPATCHED   
1358  20220116140_31     20220116140           D              DISPATCHED   
1359  20220315247_11     20220315247           S                 SUSPECT   

     incident_status_code incident_status_description investigation_status  \
0                       O                        OPEN                 Open   
1                       O                        OPEN                 Open   
2                       O                        OPEN                 Open   
3                       O                        OPEN                 Open   
4                       O                        OPEN                 Open   
...                   ...                         ...                  ...   
1355                    R        REFUSED TO COOPERATE               Closed   
1356                    R        REFUSED TO COOPERATE               Closed   
1357                    R        REFUSED TO COOPERATE               Closed   
1358                    A           CLEARED BY ARREST               Closed   
1359                    O                        OPEN                 Open   

            incident_occurred        incident_reported incident_location  ...  \
0     2022-04-07T15:00:00.000  2022-04-08T11:52:00.000        JACKSON ST  ...   
1     2022-03-18T02:30:00.000  2022-03-18T06:51:00.000        BENTON AVE  ...   
2     2022-01-18T07:45:00.000  2022-01-19T23:48:00.000     CANE RIDGE RD  ...   
3     2022-01-23T00:40:00.000  2022-01-24T06:57:00.000     BROOKWOOD TER  ...   
4     2021-12-06T12:00:00.000  2022-01-21T16:25:00.000          SOUTH ST  ...   
...                       ...                      ...               ...  ...   
1355  2022-09-13T17:00:00.000  2022-09-15T02:04:00.000         1524 1524  ...   
1356  2022-08-11T18:00:00.000  2022-08-11T22:47:00.000           930 930  ...   
1357  2022-09-27T19:16:00.000  2022-09-27T22:46:00.000           400 400  ...   
1358  2022-03-13T01:49:00.000  2022-03-13T01:49:00.000           306 306  ...   
1359  2022-06-27T12:00:00.000  2022-06-27T19:46:00.000        HARDING PL  ...   

            victim_description victim_gender victim_race victim_ethnicity  \
0     INDIVIDUAL (18 AND OVER)             U           W     Non-Hispanic   
1     INDIVIDUAL (18 AND OVER)             M           W     Non-Hispanic   
2     INDIVIDUAL (18 AND OVER)             M           B     Non-Hispanic   
3     INDIVIDUAL (18 AND OVER)             F           W     Non-Hispanic   
4     INDIVIDUAL (18 AND OVER)             M           W     Non-Hispanic   
...                        ...           ...         ...              ...   
1355  INDIVIDUAL (18 AND OVER)             M           W     Non-Hispanic   
1356  INDIVIDUAL (18 AND OVER)             F           B     Non-Hispanic   
1357  INDIVIDUAL (18 AND OVER)             F           B     Non-Hispanic   
1358  INDIVIDUAL (18 AND OVER)             F           B     Non-Hispanic   
1359  INDIVIDUAL (18 AND OVER)             F           B     Non-Hispanic   

     victim_county_resident  \
0              NON RESIDENT   
1                  RESIDENT   
2                  RESIDENT   
3                  RESIDENT   
4                  RESIDENT   
...                     ...   
1355               RESIDENT   
1356         

In [24]:
burglaries.head()

primary_key incident_number report_type report_type_description  \
0  20220167824_11     20220167824           D              DISPATCHED   
1  20220126184_31     20220126184           D              DISPATCHED   
2  20220027854_12     20220027854           D              DISPATCHED   
3  20220032825_11     20220032825           D              DISPATCHED   
4  20220010799_11     20220010799           D              DISPATCHED   

  incident_status_code incident_status_description investigation_status  \
0                    O                        OPEN                 Open   
1                    O                        OPEN                 Open   
2                    O                        OPEN                 Open   
3                    O                        OPEN                 Open   
4                    O                        OPEN                 Open   

         incident_occurred        incident_reported incident_location  ...  \
0  2022-04-07T15:00:00.000  2022-04-08T11:52:00.000        JACKSON ST  ...   
1  2022-03-18T02:30:00.000  2022-03-18T06:51:00.000        BENTON AVE  ...   
2  2022-01-18T07:45:00.000  2022-01-19T23:48:00.000     CANE RIDGE RD  ...   
3  2022-01-23T00:40:00.000  2022-01-24T06:57:00.000     BROOKWOOD TER  ...   
4  2021-12-06T12:00:00.000  2022-01-21T16:25:00.000          SOUTH ST  ...   

         victim_description victim_gender victim_race victim_ethnicity  \
0  INDIVIDUAL (18 AND OVER)             U           W     Non-Hispanic   
1  INDIVIDUAL (18 AND OVER)             M           W     Non-Hispanic   
2  INDIVIDUAL (18 AND OVER)             M           B     Non-Hispanic   
3  INDIVIDUAL (18 AND OVER)             F           W     Non-Hispanic   
4  INDIVIDUAL (18 AND OVER)             M           W     Non-Hispanic   

  victim_county_resident                                    mapped_location  \
0           NON RESIDENT   {'type': 'Point', 'coordinates': [-86.8, 36.17]}   
1               RESIDENT  {'type': 'Point', 'coordinates': [-86.77, 36.13]}   
2               RESIDENT  {'type': 'Point', 'coordinates': [-86.66, 36.04]}   
3               RESIDENT  {'type': 'Point', 'coordinates': [-86.86, 36.13]}   
4               RESIDENT  {'type': 'Point', 'coordinates': [-86.78, 36.15]}   

    rpa zone zip_code                    geometry  
0   NaN  NaN      NaN  POINT (-86.80000 36.17000)  
1  8029  817      NaN  POINT (-86.77000 36.13000)  
2   NaN  NaN      NaN  POINT (-86.66000 36.04000)  
3  5019  121      NaN  POINT (-86.86000 36.13000)  
4  6805  815      NaN  POINT (-86.78000 36.15000)  

[5 rows x 32 columns]

In [13]:
census_tracts = gpd.read_file("../data/tl_2020_47_tract.shp")
census_tracts

STATEFP COUNTYFP TRACTCE        GEOID     NAME              NAMELSAD  \
0         47      175  925200  47175925200     9252     Census Tract 9252   
1         47      175  925000  47175925000     9250     Census Tract 9250   
2         47      003  950201  47003950201  9502.01  Census Tract 9502.01   
3         47      003  950202  47003950202  9502.02  Census Tract 9502.02   
4         47      093  003300  47093003300       33       Census Tract 33   
...      ...      ...     ...          ...      ...                   ...   
1696      47      073  050501  47073050501   505.01   Census Tract 505.01   
1697      47      073  050601  47073050601   506.01   Census Tract 506.01   
1698      47      073  050602  47073050602   506.02   Census Tract 506.02   
1699      47      073  050502  47073050502   505.02   Census Tract 505.02   
1700      47      073  050503  47073050503   505.03   Census Tract 505.03   

      MTFCC FUNCSTAT      ALAND   AWATER     INTPTLAT      INTPTLON  \
0     G5020        S  227429512  1667739  +35.7438100  -085.4940227   
1     G5020        S  480712883  1225717  +35.6695378  -085.4220628   
2     G5020        S  121774227        0  +35.6517480  -086.5575518   
3     G5020        S  110617191   700793  +35.5845755  -086.5790796   
4     G5020        S    5860088   229299  +36.0020586  -083.8371218   
...     ...      ...        ...      ...          ...           ...   
1696  G5020        S   83355996        0  +36.5659159  -082.7715246   
1697  G5020        S   23847669   177932  +36.5836443  -082.6318663   
1698  G5020        S   23386546   668716  +36.5467218  -082.6458382   
1699  G5020        S   32873931   311980  +36.5332207  -082.7586081   
1700  G5020        S   17050728   568444  +36.5225397  -082.6970280   

                                               geometry  
0     POLYGON ((-85.61516 35.76106, -85.61509 35.761...  
1     POLYGON ((-85.60513 35.70854, -85.60511 35.708...  
2     POLYGON ((-86.64406 35.64029, -86.64375 35.642...  
3     POLYGON ((-86.66377 35.58189, -86.66367 35.582...  
4     POLYGON ((-83.86208 35.99255, -83.86207 35.992...  
...                                                 ...  
1696  POLYGON ((-82.85011 36.54107, -82.84955 36.541...  
1697  POLYGON ((-82.68939 36.58793, -82.68936 36.588...  
1698  POLYGON ((-82.69446 36.54856, -82.69445 36.548...  
1699  POLYGON ((-82.82242 36.52006, -82.82171 36.520...  
1700  POLYGON ((-82.73162 36.52804, -82.73108 36.528...  

[1701 rows x 13 columns]

In [14]:
burglaries = gpd.GeoDataFrame(burglaries, geometry=burglaries["geometry"], crs=census_tracts.crs)
burglaries

primary_key incident_number report_type report_type_description  \
0    20220641257_11     20220641257           W                 WITNESS   
1    20220641203_11     20220641203           W                 WITNESS   
2    20220632390_11     20220632390           W                 WITNESS   
3    20220632346_11     20220632346           W                 WITNESS   
4    20220631992_11     20220631992           W                 WITNESS   
..              ...             ...         ...                     ...   
995  20220526674_11     20220526674           S                 SUSPECT   
996  20220526664_11     20220526664           D              DISPATCHED   
997  20220526663_11     20220526663           D              DISPATCHED   
998  20220526656_11     20220526656           D              DISPATCHED   
999  20220526651_11     20220526651           D              DISPATCHED   

    incident_status_code incident_status_description investigation_status  \
0                      U                   UNFOUNDED               Closed   
1                      U                   UNFOUNDED               Closed   
2                      U                   UNFOUNDED               Closed   
3                      O                        OPEN                 Open   
4                      O                        OPEN                 Open   
..                   ...                         ...                  ...   
995                    U                   UNFOUNDED               Closed   
996                    A           CLEARED BY ARREST               Closed   
997                    O                        OPEN                 Open   
998                    O                        OPEN                 Open   
999                    O                        OPEN                 Open   

           incident_occurred        incident_reported  \
0    2022-10-14T16:00:00.000  2022-10-14T16:01:00.000   
1    2022-10-14T15:31:00.000  2022-10-14T15:31:00.000   
2    2022-10-10T16:49:00.000  2022-10-10T16:49:00.000   
3    2022-10-10T05:30:00.000  2022-10-10T16:28:00.000   
4    2022-10-09T21:00:00.000  2022-10-10T13:28:00.000   
..                       ...                      ...   
995  2022-10-19T08:10:00.000  2022-10-19T08:10:00.000   
996  2022-10-19T07:50:00.000  2022-10-19T09:37:00.000   
997  2022-10-18T13:00:00.000  2022-10-19T08:07:00.000   
998  2022-10-18T22:30:00.000  2022-10-19T08:12:00.000   
999  2022-10-17T17:00:00.000  2022-10-19T08:00:00.000   

               incident_location  ...  victim_ethnicity  \
0                      4343 4343  ...      Non-Hispanic   
1                      5800 5800  ...      Non-Hispanic   
2                        309 309  ...      Non-Hispanic   
3                QUAIL VALLEY RD  ...      Non-Hispanic   
4                   CLEARLAKE DR  ...      Non-Hispanic   
..                           ...  ...               ...   
995                    2000 2000  ...      Non-Hispanic   
996  25TH N AV & CLARKSVILLE PKE  ...      Non-Hispanic   
997                  COMMERCE ST  ...               NaN   
998                  HARMONY WAY  ...               NaN   
999                   HARDING PL  ...               NaN   

     victim_county_resident  \
0                  RESIDENT   
1                  RESIDENT   
2                  RESIDENT   
3                  RESIDENT   
4              NON RESIDENT   
..                      ...   
995                RESIDENT   
996                RESIDENT   
997                     NaN   
998                     NaN   
999                     NaN   

                                       mapped_location  \
0    {'type': 'Point', 'coordinates': [-86.612, 36....   
1    {'type': 'Point', 'coordinates': [-86.588, 36....   
2    {'type': 'Point', 'coordinates': [-86.582, 36....   
3    {'type': 'Point', 'coordinates': [-86.64, 36.14]}   
4     {'type': 'Point', 'coordinates': [-86.6, 36.05]}   
..                                                 ...   
995  {'ty

1: only 1000 rows ok?
2: discuss lambda, Point, x, .apply.
3: discuss GeoDataFrame, crs.
4: Time in groups tonight?
5: Help with branch to push...can't get it to push.
6:

#MPD dataset API
# Load API Key into dictionary: credentials
with open('app_token.json') as fi:
    credentials = json.load(fi)
# Returns value associated with dictionary key
api_key = credentials['token']
#List of dates of interest
dates = ("'2022-01-01T00:00:00'", "'2022-09-30T23:59:59'")
#NIBRS codes for Aggrivated burglary
nibrs = ['220', '13A', '120', '520']

#2020 American Community Survey API Calls
# Load API Key into dictionary: credentials
with open('census_key.json') as fi:
    credentials = json.load(fi)
# Returns value associated with dictionary key
census_api = credentials['api_key']
#FPIS codes for Tennessee, davidson county
fpis= ['47', '37']
#Datasets of interest
Census_Datasets = ('NAME','B01001_001E','S1901_C01_012E')

#formats response to .json()
income = income.json()
# creates pandas dataframe from res variable (response.json)
Med_Household_income_df = pd.DataFrame(income)
header = Med_Household_income_df.iloc[0]
Med_Household_income_df = Med_Household_income_df[1:]
Med_Household_income_df.columns = header
Med_Household_income_df = Med_Household_income_df.rename(columns={'S1901_C01_012E': 'med_household_income'})
Med_Household_income_df

#formats response to .json()
res = response.json()
# creates pandas dataframe from res variable (response.json)
population_df = pd.DataFrame(res)
header = population_df.iloc[0]
population_df = population_df[1:]
population_df.columns = header
population_df = population_df.rename(columns={'S0101_C01_001E': 'population'})
population_df
